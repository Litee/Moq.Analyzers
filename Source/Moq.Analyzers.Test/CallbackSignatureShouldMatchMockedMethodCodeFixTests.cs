using ApprovalTests;
using ApprovalTests.Reporters;
using Microsoft.CodeAnalysis.CodeFixes;
using Microsoft.CodeAnalysis.Diagnostics;
using System.IO;
using TestHelper;
using Xunit;

namespace Moq.Analyzers.Test
{
    [UseReporter(typeof(DiffReporter))]
    public class CallbackSignatureShouldMatchMockedMethodCodeFixTests : CodeFixVerifier
    {

        [Fact]
        public void ShouldSuggestQuickFixIfBadParameters()
        {
            Approvals.Verify(VerifyCSharpFix(File.ReadAllText("Data/CallbackSignatureShouldMatchMockedMethod.cs")));
        }

        protected override CodeFixProvider GetCSharpCodeFixProvider()
        {
            return new CallbackSignatureShouldMatchMockedMethodCodeFix();
        }

        protected override DiagnosticAnalyzer GetCSharpDiagnosticAnalyzer()
        {
            return new CallbackSignatureShouldMatchMockedMethodAnalyzer();
        }
    }
}