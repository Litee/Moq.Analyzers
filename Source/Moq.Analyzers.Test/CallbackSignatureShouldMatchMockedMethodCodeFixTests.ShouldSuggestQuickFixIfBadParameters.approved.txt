using Moq;
using System;
using System.Collections.Generic;

namespace CallbackSignatureShouldMatchMockedMethod
{
    interface IFoo
    {
        int Do(string s);

        int Do(int i, string s, DateTime dt);

        int Do(List<string> l);
    }

    class MyUnitTests
    {
        void TestBadCallbacks()
        {
            var mock = new Mock<IFoo>();
            mock.Setup(x => x.Do(It.IsAny<string>())).Callback((string s) => { });
            mock.Setup(x => x.Do(It.IsAny<string>())).Callback((string s) => { });
            mock.Setup(x => x.Do(It.IsAny<int>(), It.IsAny<string>(), It.IsAny<DateTime>())).Callback((int i, string s, DateTime dt) => { });
            mock.Setup(x => x.Do(It.IsAny<List<string>>())).Callback((List<string> l) => { });
        }

        void TestGoodSetupAndParameterlessCallback()
        {
            var mock = new Mock<IFoo>();
            mock.Setup(x => x.Do(It.IsAny<string>())).Callback(() => { });
            mock.Setup(x => x.Do(It.IsAny<int>(), It.IsAny<string>(), It.IsAny<DateTime>())).Callback(() => { });
            mock.Setup(x => x.Do(It.IsAny<List<string>>())).Callback(() => { });
        }

        void TestGoodSetupAndCallback()
        {
            var mock = new Mock<IFoo>();
            mock.Setup(x => x.Do(It.IsAny<string>())).Callback((string s) => { });
            mock.Setup(x => x.Do(It.IsAny<int>(), It.IsAny<string>(), It.IsAny<DateTime>())).Callback((int i, string s, DateTime dt) => { });
            mock.Setup(x => x.Do(It.IsAny<List<string>>())).Callback((List<string> l) => { });
        }

        void TestGoodSetupAndReturnsAndCallback()
        {
            var mock = new Mock<IFoo>();
            mock.Setup(x => x.Do(It.IsAny<string>())).Returns(0).Callback((string s) => { });
            mock.Setup(x => x.Do(It.IsAny<int>(), It.IsAny<string>(), It.IsAny<DateTime>())).Returns(0).Callback((int i, string s, DateTime dt) => { });
            mock.Setup(x => x.Do(It.IsAny<List<string>>())).Returns(0).Callback((List<string> l) => { });
        }

        void MyGoodSetupAndReturns()
        {
            var mock = new Mock<IFoo>();
            mock.Setup(x => x.Do(It.IsAny<string>())).Returns((string s) => { return 0; });
            mock.Setup(x => x.Do(It.IsAny<int>(), It.IsAny<string>(), It.IsAny<DateTime>())).Returns((int i, string s, DateTime dt) => { return 0; });
            mock.Setup(x => x.Do(It.IsAny<List<string>>())).Returns((List<string> l) => { return 0; });
        }
    }
}